### ISSUE TITLE
RFC: inline recipe bootstrap

### ISSUE THREAD

**Issue by kallisti5** [haikuports/haikuporter#134]
We currently keep a separate bootstrap repository. This repository has recipes that have been stripped down the bare minimum requirements to bootstrap Haiku.  One problem with this design is since bootstraps are rarely required (but important!) the packages in the bootstrap repository are normally in dire-shape and outdated.

It might make sense to bring the minimum bootstrap logic into the normal recipes.

Example:
```
SUMMARY="a package full of lols"
DESCRIPTION=this is a package full of all the lols"

PROVIDES="
    lols = $portVersion
    extreme_lols = $portVersion
"

REQUIRES="
   haiku
   extreme_stuff
"
.
.
BUILD()
{
    make all the things
}
.
.
# Minimum package for bootstrap
MINIMUM_PROVIDES="
    lols
"
MINIMUM_REQUIRES="
    haiku
"
MINIMUM_BUILD()
{
    make only a few things
}
```

The MINIMUM_X stuff isn't required, but when a flag is toggled in haikuporter, the MINIMUM sections are used instead.

This allows us to:
  * Tie together common things that make sense (package URLs)
  * Help ensure bootstrap recipes are maintained

There are likely better ways to design this process, i'm just giving a high level idea looking for feedback :-)

**Comment by kallisti5** [comment]
```
./haikuporter --bootstrap   lols  |  haikuporter --minimum lols    etc..
<build stuff>
lols_bootstrap-1.2.4.hpkg
```

**Comment by pulkomandy** [comment]
Expected result:
- The bootstrap part is still not upated by anyone
- People make changes to the non-bootstrap part which breaks the bootstrap even more

Sure this is a good idea?

What we need is a buildbot that does a bootstrap everynight, so we noticed when things are broken. Then maybe people will start caring about it.

Also, the recipes for bootstrap are in some case completely different than the normal ones, and usually more complex. It is not always an easy feat to do a cross-build of things. For example in the case of Python, you should use a special version of the sources which was heavily patched to allow it (or at least that's how it was done last time I looked into it).

Things I remember from latest discussions about bootstrapping (that was a few years ago):
- We should reduce the bootstrap repo to a bare minimum list of packages
- To achieve this, we should make a minimal haiku bootstrap package: no translators (removes libjpeg libpng ...), no app_server (removes freetype), no locale kit (removes ICU), etc. Basically just the kernel and a consoled or telnet server + bash.
- From there on, we could run the bootstrap from inside haiku, which means we would use native recipes for all the remaining dependencies.

**Comment by kallisti5** [comment]
Reducing the repo would help, but I don't think it is a solution.

We currently have 31 recipes. Removing icu and freetype are pretty much the only ones that could be removed. (there is no libjpeg, libpng, etc in haikuports.cross)


> The bootstrap part is still not upated by anyone

That's not an excuse. This change would ensure the minimum/bootstrap logic would need to be updated. This allows the "minimum package logic" to piggyback on some of the non-minimum logic which means more of it *will* be updated even if the porter gets lazy (which we should call out via the commit ML). (URL's, SHA checksums, descriptions, licenses, etc).  It also encourages people to keep up with the minimum package logic.

> People make changes to the non-bootstrap part which breaks the bootstrap even more
Since the minimum package logic is isolated, changes to the non-mininum recipe logic should be insulated from the minimum logic. This comment also doesn't make a lot of sense because the only potential breakage is if the "building steps" change in some drastic way.

**Comment by pulkomandy** [comment]
I don't think you want the URL and SHA to be changed without checks. It's unlikely that the recipe would still work, so whenever you want to do a bootstrap, the first step would be "alright, when did this built properly for the last time and which version and URL were used back then". Which is why they were put in a separate branch in the first place: we need to keep track of what worked last time.

It should not be a problem if these recipes are out of date. As long as you get a running Haiku in the end, you can then continue the bootstrap process from there with more up to date recipes. We don't want the bootstrap recipes to follow and match the normal ones, we want them to just work and stay frozen as much as possible. They are just used to get an Haiku system up and running, but then, the second stage of the bootstrap starts, and everything is rebuilt.

**Comment by avongluck-r1soft** [comment]
> I don't think you want the URL and SHA to be changed without checks. It's unlikely that the recipe would still work, so whenever you want to do a bootstrap, the first step would be "alright, when did this built properly for the last time and which version and URL were used back then". Which is why they were put in a separate branch in the first place: we need to keep track of what worked last time.

Nobody ever said the URL and SHA would be changed without checks. The minimum build of a recipe and the standard build share the same files/revisions/etc.

> It should not be a problem if these recipes are out of date. As long as you get a running Haiku in the end, you can then continue the bootstrap process from there with more up to date recipes. We don't want the bootstrap recipes to follow and match the normal ones, we want them to just work and stay frozen as much as possible. They are just used to get an Haiku system up and running, but then, the second stage of the bootstrap starts, and everything is rebuilt.

It's less about having an occasional bootstrap recipe out of date, and more about keeping at least some of them up to date. Automated testing of minimal package recipies gets easier when all we need to do is haikuporter --minimum-variant   instead of re-cloning a new recipe repository.   Going through the bootstrap process a few months ago, pretty much every recipe was broken in some way. (Broken download links, broken mirror links, issues resolved long ago, etc)

> If we try to get the bootstrap to match the versions used for normal recipes, then they will always be broken and no one will ever be able to do a bootstrap (by the time one fixes all recipes, a dozen updates are merged already and you need to restart the work).

As I understand it, most of the changes around the bootstrap recipes are disabling extra features through configuration scripts. From what I can tell looking at the patches in the haikuports.cross repository, we never really did any bootstrap-specific patching.

>Moreover, it introduces more special-casing for no reason in the recipes, which means haikuporter would need to behave differently in bootstrap mode. This should not be needed, the only difference would be the host system and compiler used, but that should be transparent and haikuporter should process recipes exactly the same way in bootstrap and non-bootstrap mode.

--variant-minimum (or whatever) would only do the following:
   * Use MINIMUM_XXX stanzas if available
   * Prefix package names with minimum/bootstrap/etc

This is our (young) ecosystem, there are a lot of various paths we could take to improve this situation.
I'm not sure wh we dig our heels in so often without willingness to experiment (especially on things which  won't block R1B1)

**Comment by pulkomandy** [comment]
We didn't do much patching, but as you noticed, the resulting bootstrap process simply doesn't work. Unless you use the exact same version of OpenSUSE it was run on at the time.

For example, cross compiling Python is a mess and does not even use upstream sources but an heavily patched version: http://www.trevorbowen.com/2013/10/07/cross-compiling-python-2-7-5-for-embedded-linux/

Otherwise you end up with weird buld failures unless you use the exact same setup as the host python you are running (same directory layout, etc). when I fixed Python bootstrap to work under Debian, it broke on OpenSUSE because of this.

**Comment by jessicah** [comment]
I think the first step is getting the bootstrap process running well again.

Probably the best way to do this is to use a container (e.g. docker) to run bootstrap builds. This will give us a 100% reproducible environment that devs can run on their systems that differ to what's known tested to work in the docker image. We could even use OpenSuSE in our docker image, if that works best... icky :p

**Comment by korli** [comment]
To bootstrap x86_64, I used a ubuntu x86_64 16.04 installation to cross-build the haiku-bootstrap image. Then in the do-bootstrap step, there were circular dependencies related to the use of ca_root_certificates (any architecture), which doesn't seem correctly handled. I just commented these dependencies to solve this minor issue.
